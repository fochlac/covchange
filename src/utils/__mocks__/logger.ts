export default jest.fn(console.log)
