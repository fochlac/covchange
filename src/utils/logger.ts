import { createWriteStream, existsSync, mkdirSync } from 'fs'

/**
 *   Log Levels:
 *   0: no error, very important info
 *   1: error: critical error in app
 *   2: error: internal error in app
 *   3: error: minor error
 *
 *   4: info: important info (bad request/login)
 *   5: info: minor info
 *
 *   6: trace: tracing informations
 *   7: unimportant tracing information
 *
 *  10: full trace: (requests objects, etc)
 *
 **/
let currentDate = getDate().day
if (!existsSync(global.appRoot + 'log')) {
	mkdirSync(global.appRoot + 'log')
}

let logStream = createWriteStream(global.appRoot + 'log/output' + getDate().day + '.txt', {
	flags: 'a',
	encoding: 'utf8',
	autoClose: true,
})

logStream.on('error', err => {
	console.log(err)
})

export default (level: number, ...message: any[]): void => {
	const now = getDate()
	const messages = message.map(parseMessage).join(' ')
	const logMessage = `${now.day} - ${now.time} - ${level} - ${messages}`

	if (level <= global.logLevel) {
		console.log(logMessage)
	}

	if (currentDate !== getDate().day) {
		currentDate = getDate().day
		logStream.end()
		logStream = createWriteStream(global.appRoot + 'log/output' + getDate().day + '.txt', {
			flags: 'a',
			encoding: 'utf8',
			autoClose: true,
		})
	}

	logStream.write(logMessage)
}

function parseMessage(item) {
	let output
	if (['string', 'number', 'boolean'].includes(typeof item)) {
		output = item
	} else {
		try {
			if (Object.prototype.toString.call(item) === '[object Error]') {
				output = '\n' + item.stack.toString()
			} else {
				output = '\n' + JSON.stringify(item)
			}
		} catch (err) {
			console.log(err + '\n' + item)
			output = ''
		}
	}
	return output
}

function getDate() {
	const date = new Date()
	return {
		day: date.getFullYear() + '_' + date.getMonth() + '_' + date.getDate(),
		time: date.getHours() + ':' + date.getMinutes() + ':' + date.getSeconds(),
	}
}
