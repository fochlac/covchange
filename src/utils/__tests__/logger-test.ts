import '../../__tests__/mock-config'

// @ts-ignore
import { createWriteStream, existsSync, mkdirSync, mockStream } from 'fs'

import { useFakeTimers } from 'sinon'

const clock = useFakeTimers({ now: new Date(1234567891011), shouldAdvanceTime: true, advanceTimeDelta: 50 })

jest.mock('fs', () => {
	const mockStream = { end: jest.fn(), write: jest.fn(), on: jest.fn() }
	return {
		existsSync: jest.fn(() => false),
		mkdirSync: jest.fn(),
		createWriteStream: jest.fn(() => mockStream),
		mockStream,
	}
})

const error = new Error('test')
error.stack = 'Error: test'

describe('logger', () => {
	beforeEach(() => {
		jest.spyOn(global.console, 'log').mockImplementation(() => null)
	})

	afterEach(() => {
		mockStream.end.mockClear()
		mockStream.write.mockClear()
		mockStream.on.mockClear()
		// @ts-ignore
		existsSync.mockClear()
		// @ts-ignore
		mkdirSync.mockClear()
		// @ts-ignore
		createWriteStream.mockClear()
		// @ts-ignore
		global.console.log.mockRestore()
	})

	it('create dir if missing ,write info to the output and console log', () => {
		const logger = require('../logger').default
		logger(1, 'testmessage1')
		expect(existsSync).toBeCalledTimes(1)
		expect(existsSync).toBeCalledWith('/test/root/log')
		expect(mkdirSync).toBeCalledTimes(1)
		expect(mkdirSync).toBeCalledWith('/test/root/log')
		expect(createWriteStream).toBeCalledTimes(1)
		// @ts-ignore
		expect(createWriteStream.mock.calls).toMatchSnapshot()
		expect(mockStream.write).toBeCalledTimes(1)
		expect(mockStream.write.mock.calls).toMatchSnapshot()
		expect(console.log).toBeCalledTimes(1)
		// @ts-ignore
		expect(console.log.mock.calls).toMatchSnapshot()
	})

	it('should parse errors , numbers and objects', () => {
		const logger = require('../logger').default
		logger(1, 'testmessage1', 1, error, { test: 'test' })
		expect(existsSync).toBeCalledTimes(0)
		expect(mkdirSync).toBeCalledTimes(0)
		expect(createWriteStream).toBeCalledTimes(0)
		expect(mockStream.write).toBeCalledTimes(1)
		expect(mockStream.write.mock.calls).toMatchSnapshot()
		expect(console.log).toBeCalledTimes(1)
		// @ts-ignore
		expect(console.log.mock.calls).toMatchSnapshot()
	})

	it('should parse errors , numbers and objects and pass on errors on failure', () => {
		const logger = require('../logger').default
		const testObject: { test: string; testObject?: any } = { test: 'test' }
		testObject.testObject = testObject
		logger(1, 'testmessage1', 1, error, { test: 'test' }, testObject)
		expect(existsSync).toBeCalledTimes(0)
		expect(mkdirSync).toBeCalledTimes(0)
		expect(createWriteStream).toBeCalledTimes(0)
		expect(mockStream.write).toBeCalledTimes(1)
		expect(mockStream.write.mock.calls).toMatchSnapshot()
		expect(console.log).toBeCalledTimes(2)
		// @ts-ignore
		expect(console.log.mock.calls).toMatchSnapshot()
	})

	it('should create new file if day changes', () => {
		const logger = require('../logger').default
		logger(1, 'testmessage1')
		expect(createWriteStream).toBeCalledTimes(0)
		expect(mockStream.end).toBeCalledTimes(0)
		expect(mockStream.write).toBeCalledTimes(1)

		clock.tick(1000 * 60 * 60 * 30)
		logger(1, 'testmessage1')
		expect(createWriteStream).toBeCalledTimes(1)
		expect(mockStream.end).toBeCalledTimes(1)
		expect(mockStream.write).toBeCalledTimes(2)
	})

	it('should output error in console on write error', () => {
		jest.resetModules()
		const { createWriteStream, mockStream } = require('fs')
		require('../logger').default
		expect(createWriteStream).toBeCalledTimes(1)
		expect(mockStream.on).toBeCalledTimes(1)
		mockStream.on.mock.calls[0][1]('testerror')
		expect(console.log).toBeCalledTimes(1)
		// @ts-ignore
		expect(console.log).toBeCalledWith('testerror')
	})

	it('write info to the output but not to console.log', () => {
		let logger
		// @ts-ignore
		jest.mock('fs', () => {
			const mockStream = { end: jest.fn(), write: jest.fn(), on: jest.fn() }
			return {
				existsSync: jest.fn(() => true),
				mkdirSync: jest.fn(),
				createWriteStream: jest.fn(() => mockStream),
				mockStream,
			}
		})
		jest.resetModules()
		logger = require('../logger').default
		const { createWriteStream, existsSync, mkdirSync, mockStream } = require('fs')
		logger(10, 'testmessage1')
		expect(existsSync).toBeCalledTimes(1)
		expect(existsSync).toBeCalledWith('/test/root/log')
		expect(mkdirSync).toBeCalledTimes(0)
		expect(createWriteStream).toBeCalledTimes(1)
		// @ts-ignore
		expect(createWriteStream.mock.calls).toMatchSnapshot()
		expect(mockStream.write).toBeCalledTimes(1)
		expect(mockStream.write.mock.calls).toMatchSnapshot()
		expect(console.log).toBeCalledTimes(0)
	})
})
