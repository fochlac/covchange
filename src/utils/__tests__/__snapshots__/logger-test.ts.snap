// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logger create dir if missing ,write info to the output and console log 1`] = `
Array [
  Array [
    "/test/root/log/output2009_1_14.txt",
    Object {
      "autoClose": true,
      "encoding": "utf8",
      "flags": "a",
    },
  ],
]
`;

exports[`logger create dir if missing ,write info to the output and console log 2`] = `
Array [
  Array [
    "2009_1_14 - 0:31:31 - 1 - testmessage1",
  ],
]
`;

exports[`logger create dir if missing ,write info to the output and console log 3`] = `
Array [
  Array [
    "2009_1_14 - 0:31:31 - 1 - testmessage1",
  ],
]
`;

exports[`logger should parse errors , numbers and objects 1`] = `
Array [
  Array [
    "2009_1_14 - 0:31:31 - 1 - testmessage1 1 
Error: test
    at Object.it (C:\\\\gitHub\\\\covwatcher\\\\src\\\\utils\\\\__tests__\\\\logger-test.ts:58:32)
    at Object.asyncJestTest (C:\\\\gitHub\\\\covwatcher\\\\node_modules\\\\jest-jasmine2\\\\build\\\\jasmineAsyncInstall.js:102:37)
    at resolve (C:\\\\gitHub\\\\covwatcher\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:43:12)
    at new Promise (<anonymous>)
    at mapper (C:\\\\gitHub\\\\covwatcher\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:26:19)
    at promise.then (C:\\\\gitHub\\\\covwatcher\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:73:41)
    at process._tickCallback (internal/process/next_tick.js:68:7) 
{\\"test\\":\\"test\\"}",
  ],
]
`;

exports[`logger should parse errors , numbers and objects 2`] = `
Array [
  Array [
    "2009_1_14 - 0:31:31 - 1 - testmessage1 1 
Error: test
    at Object.it (C:\\\\gitHub\\\\covwatcher\\\\src\\\\utils\\\\__tests__\\\\logger-test.ts:58:32)
    at Object.asyncJestTest (C:\\\\gitHub\\\\covwatcher\\\\node_modules\\\\jest-jasmine2\\\\build\\\\jasmineAsyncInstall.js:102:37)
    at resolve (C:\\\\gitHub\\\\covwatcher\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:43:12)
    at new Promise (<anonymous>)
    at mapper (C:\\\\gitHub\\\\covwatcher\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:26:19)
    at promise.then (C:\\\\gitHub\\\\covwatcher\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:73:41)
    at process._tickCallback (internal/process/next_tick.js:68:7) 
{\\"test\\":\\"test\\"}",
  ],
]
`;

exports[`logger should parse errors , numbers and objects and pass on errors on failure 1`] = `
Array [
  Array [
    "2009_1_14 - 0:31:31 - 1 - testmessage1 1 
Error: test
    at Object.it (C:\\\\gitHub\\\\covwatcher\\\\src\\\\utils\\\\__tests__\\\\logger-test.ts:73:32)
    at Object.asyncJestTest (C:\\\\gitHub\\\\covwatcher\\\\node_modules\\\\jest-jasmine2\\\\build\\\\jasmineAsyncInstall.js:102:37)
    at resolve (C:\\\\gitHub\\\\covwatcher\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:43:12)
    at new Promise (<anonymous>)
    at mapper (C:\\\\gitHub\\\\covwatcher\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:26:19)
    at promise.then (C:\\\\gitHub\\\\covwatcher\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:73:41)
    at process._tickCallback (internal/process/next_tick.js:68:7) 
{\\"test\\":\\"test\\"} ",
  ],
]
`;

exports[`logger should parse errors , numbers and objects and pass on errors on failure 2`] = `
Array [
  Array [
    "TypeError: Converting circular structure to JSON
[object Object]",
  ],
  Array [
    "2009_1_14 - 0:31:31 - 1 - testmessage1 1 
Error: test
    at Object.it (C:\\\\gitHub\\\\covwatcher\\\\src\\\\utils\\\\__tests__\\\\logger-test.ts:73:32)
    at Object.asyncJestTest (C:\\\\gitHub\\\\covwatcher\\\\node_modules\\\\jest-jasmine2\\\\build\\\\jasmineAsyncInstall.js:102:37)
    at resolve (C:\\\\gitHub\\\\covwatcher\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:43:12)
    at new Promise (<anonymous>)
    at mapper (C:\\\\gitHub\\\\covwatcher\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:26:19)
    at promise.then (C:\\\\gitHub\\\\covwatcher\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:73:41)
    at process._tickCallback (internal/process/next_tick.js:68:7) 
{\\"test\\":\\"test\\"} ",
  ],
]
`;

exports[`logger write info to the output but not to console.log 1`] = `
Array [
  Array [
    "/test/root/log/output2009_1_15.txt",
    Object {
      "autoClose": true,
      "encoding": "utf8",
      "flags": "a",
    },
  ],
]
`;

exports[`logger write info to the output but not to console.log 2`] = `
Array [
  Array [
    "2009_1_15 - 6:31:31 - 10 - testmessage1",
  ],
]
`;
